encfs/CipherFileIO.cpp

```diff
@@ -483,36 +483,92 @@ ssize_t CipherFileIO::writeOneBlock(const IORequest &req) {
     return -EPERM;
   }
 
+  int cbs = fsConfig->cipher->cipherBlockSize();
   unsigned int bs = blockSize();
   off_t blockNum = req.offset / bs;
 
-  if (haveHeader && fileIV == 0) {
+  if (haveHeader && (fileIV == 0)) {
     int res = initHeader();
     if (res < 0) {
       return res;
     }
   }
 
   bool ok;
-  if (req.dataLen != bs) {
-    ok = streamWrite(req.data, (int)req.dataLen,
-                     blockNum ^ fileIV);  // cast works because we work on a
-                                          // block and blocksize fit an int
-  } else {
-    ok = blockWrite(req.data, (int)req.dataLen,
-                    blockNum ^ fileIV);  // cast works because we work on a
-                                         // block and blocksize fit an int
+
+  int padBytes = 0;
+  if (haveCBCPadding) {
+    if (!fsConfig->reverseEncryption) {
+      // we need padding bytes so that dataLen is a multiple of cipherBlockSize()
+      padBytes = cbs - (req.dataLen % cbs);
+      // we then add the first padding byte : 0x80
+      req.data[req.dataLen] = 0x80;
+      // if needed, we add the other padding bytes : 0x00
+      memset(req.data + req.dataLen + 1, 0x00, padBytes - 1);
+      ok = blockWrite(req.data, (int)req.dataLen + padBytes,
+                      blockNum ^ fileIV);  // cast works because we work on a
+                                           // block and blocksize fit an int
```

Comment:
On error, please return early to cut down "if maze". In other words,

    if(!ok) {
    log error
    return
    }